### PR TITLE
[JENKINS-59451] Prevent unclosed file handlers in Slave 2 Master security Admin Whitelist rule AND in agent standard output swapper

### PR DESCRIPTION
**Pull request by StefanSpieker**
See [JENKINS-59451](https://issues.jenkins-ci.org/browse/JENKINS-59451).

Fixed another Spotbugs type: [unsatisfied-obligation](https://spotbugs.readthedocs.io/en/latest/bugDescriptions.html#obl-method-may-fail-to-clean-up-stream-or-resource-obl-unsatisfied-obligation)

### Proposed changelog entries

* Internal: Spotbugs issues fixed

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### CODE REVIEW

**Review by rsandell — APPROVED** [review]

**Review by fcojfernandez — APPROVED** [review]

**Comment by rsandell** [comment]
Maybe need to rerun CI

**Comment by StefanSpieker** [comment]
@rsandell: How can I trigger the CI to rerun?

**Comment by fcojfernandez** [comment]
@StefanSpieker close and re-open the PR

**Review by oleg-nenashev — APPROVED** [review]
LGTM. It looks like it fixes real defects, so we may need to create follow-up issues

**Comment by oleg-nenashev** [comment]
retriggering CI

File: core/src/main/java/jenkins/security/s2m/AdminWhitelistRule.java [hunk 1]
```diff
@@ -65,12 +65,16 @@ public AdminWhitelistRule() throws IOException, InterruptedException {
...
         // 0-byte file is used as a signal from the admin to prevent this overwriting
-        placeDefaultRule(
-                new File(jenkins.getRootDir(), "secrets/whitelisted-callables.d/default.conf"),
-                getClass().getResourceAsStream("callable.conf"));
-        placeDefaultRule(
-                new File(jenkins.getRootDir(), "secrets/filepath-filters.d/30-default.conf"),
-                transformForWindows(getClass().getResourceAsStream("filepath-filter.conf")));
+        try (InputStream collableConf = getClass().getResourceAsStream("callable.conf")) {
+            placeDefaultRule(
+                    new File(jenkins.getRootDir(), "secrets/whitelisted-callables.d/default.conf"),
```
[the remark this answers is not in the document]
Comment (StefanSpieker) [reply core/src/main/java/jenkins/security/s2m/AdminWhitelistRule.java]: java.io.File doesn't represent an open file, it represents a path in the filesystem.
https://stackoverflow.com/questions/4752266/why-java-io-file-doesnt-have-a-close-method

So no need to take care of the file

File: core/src/main/java/jenkins/security/s2m/AdminWhitelistRule.java [hunk 2]
```diff
@@ -65,12 +65,16 @@ public AdminWhitelistRule() throws IOException, InterruptedException {
...
+                    collableConf);
+        }
+        try (InputStream filepathFilter = getClass().getResourceAsStream("filepath-filter.conf")) {
+            placeDefaultRule(
+                    new File(jenkins.getRootDir(), "secrets/filepath-filters.d/30-default.conf"),
```
[the remark this answers is not in the document]
Comment (StefanSpieker) [reply core/src/main/java/jenkins/security/s2m/AdminWhitelistRule.java]: see above comment.

**Comment by oleg-nenashev** [comment]
@StefanSpieker while I am trying to get CI passing here, could you please crate a bug in Jenkins JIRA for it? Can be a part of the SpotBugs cleanup EPIC, but I think we might want to backport it

**Comment by StefanSpieker** [comment]
I created a task for that: https://issues.jenkins-ci.org/browse/JENKINS-59451

**Comment by oleg-nenashev** [comment]
Looks like the test failure is real. I suspect `StandardOutputSwapper` closes something unexpected

**Comment by StefanSpieker** [comment]
It is quite hard for me to find the really failing test. Have you thought abot integrating the warnings-ng plugin and collecting the results?

I guess the plugin is not installed, but if it would you could add this after the build. At least we are using this in all our jenkins projects and it is quite helpful:

`recordIssues enabledForFailure: true, tools: [mavenConsole(), java(), javaDoc()]`

**Comment by oleg-nenashev** [comment]
> It is quite hard for me to find the really failing test. Have you thought abot integrating the warnings-ng plugin and collecting the results?

There is https://github.com/jenkinsci/pom/pull/43 by @tomasbjerre  which probably addresses it

**Comment by StefanSpieker** [comment]
I have problems getting the tests run on my machine at all. All tests that involve the    

>  @Rule
>     public JenkinsRule j = new JenkinsRule();

  does not seem to work on my windows 10 machine, at least not within intelliJ.

Have you an idea how to get these running on Windows?

**Comment by timja** [comment]
> I have problems getting the tests run on my machine at all. All tests that involve the
> 
> > @rule
> > public JenkinsRule j = new JenkinsRule();
> 
> does not seem to work on my windows 10 machine, at least not within intelliJ.
> 
> Have you an idea how to get these running on Windows?

what's the error?

**Comment by StefanSpieker** [comment]
@timja: Actually it seems these were missing priviledges and a locale related issue, for what I opened another PR. At least now the tests pass locally.

**Comment by StefanSpieker** [comment]
@varyvol actually all tests pass on my system. And I'm not sure which test fails on CI/CD because it seems to time out. Can you help me in finding out which tests failed?

**Comment by timja** [comment]
`startupShouldNotFailOnExceptionOnlineListener`
they appear to be in the build log?

**Comment by StefanSpieker** [comment]
This test passes locally. Either with all tests running via mvn test or with just this single test from the IDE.

**Comment by varyvol** [comment]
Maybe it was a temporary infra issue. I'm going to close and reopen to relaunch CI.

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
We will need to fix CI before merging

**Comment by varyvol** [comment]
@StefanSpieker could you maybe dedicate some time to review and fix the test failure? This seems close to be ready once that issue is solved.

**Comment by StefanSpieker** [comment]
Actually it is really hard to find out for me what the real problem is. Somewhere in jenkins-test. The Errors I get are always:

> jenkins.util.io.CompositeIOException: Unable to delete 'C:\Projekte\jenkins\jenkins\test\target\j h6629989225187982938'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.

I guess, now Jenkins does not release everything. So instead of fixing an unclosed file handlers, it seems I created new ones...

**Comment by alecharp** [comment]
@StefanSpieker could you have a look at the build error and check whether or not it's caused by the changeset of this PR? Thanks.

**Comment by MRamonLeon** [comment]
Proposed for close because there is no progress since a long time

**Comment by rsandell** [comment]
Close since no opposition for 14 days.

**Pull request closed without merging**